### ISSUE TITLE
Kong proxy gets stuck after surge in requests, responding only with HTTP status code 404

### ISSUE THREAD

**Issue by seh** [Kong/kong#8152]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version

Kong 2.6.0

### Current Behavior

Using Kong Ingress Controller version 2.0.6 with proxy version 2.6.0 (with the same behavior observed with versions 2.5.0 and 2.5.1), when subjected to surges in requests, the proxy starts responding to all requests with HTTP status code 404, with the following message body:
```json
{"message":"no Route matched with those values"}
```
The number of open file descriptors in the proxy container rises sharply around this time—probably beforehand—and the CPU usage falls dramatically once the proxy settles into this mode.

In some cases, as this surge in open file descriptors climbs, the proxy container stops responding to the kubelet's issued liveness probes. When enough of these probes time out, the kubelet kills the proxy container (first via SIGQUIT—per the _Dockerfile_'s "STOPSIGNAL" field—then later via SIGTERM). The replacement container fares better for a while, responding normally to requests.

We don't see any complaints noted in Kong's log output. As configured, Kong sends both its admin and proxy error and access logs to stderr and stdout respectively. Kubernetes collects both streams together. 

When we create new Kong pods on the same Kubernetes node as one of these "stuck" Kong pods, the new pods work fine, suggesting that there isn't some machine-level exhaustion or poisoning.

I first brought this problem up for discussion [in the "kong" channel of the "Kubernetes" Slack workspace](https://kubernetes.slack.com/archives/CDCA87FRD/p1638827090108800).

### Expected Behavior

The Kong proxy should endure this surge in requests, and either proxy them as well as the upstream servers can tolerate, or respond with an appropriate HTTP status code indicating that it's too busy at the moment. It should not respond with status code 404 as though there are no known resources available for all the requests' URLs.

### Steps To Reproduce

1. We notice this occurring more within AWS EC2 than within GCP (in GKE-managed Kubernetes clusters).
1. Run a set of KIC pods, each with one container for the ingress controller and one for the proxy.  
  We request two CPUs and 4 GiB of memory for the proxy container, and .1 CPUs and 128 MiB of memory for the ingress controller container.
1. Activate the Prometheus plugin and scrape Kong's pods to collect metrics.
1. Situate these Kong servers behind a Kubernetes _Service_ of type "LoadBalancer," with external traffic policy "Cluster."
1. Configure the _Service_ to summon an EC2 Network Load Balancer to arrange network ingress.
1. Use several _Ingress_ objects to define services and routes within Kong's proxies.
1. Subject these servers to a moderate request load (~300 RPS).  
  Assume that most of these requests deserve a response with HTTP status code 200.
1. Monitor the number of file descriptors open in the "proxy" container.  
  A PromQL expression like `container_file_descriptors{namespace="kong", pod=~"ingress-kong.+", container="proxy"}` suffices.
1. Subject these servers to an increased request load (~500 RPS).
1. Observe the file descriptor counts increasing rapidly.  
  These usually quadruple to reach as many as 8,000 over the course of ten minutes.
1. Watch the events and container status related to the "proxy" containers, to see if liveness probes are failing or if the kubelet restarts the containers.
1. Watch the HTTP status codes in Kong's responses.  
  At some point, they'll all change to 404.
1. Watch the CPU usage of the proxy containers.  
  At some point, they'll fall dramatically, coincident with Kong responding with the 404 status code.
1. Restart the NGINX worker processes in one of these "stuck" proxy containers.
1. Observe that Kong starts serving requests normally again.

### Anything else?

**Operating system:** Flatcar Container Linux version 2765.1.0, kernel version 5.10.16-flatcar
**Kong configuration:**
<details>
<summary>HTTP GET against Admin port</summary>

```json
{
  "configuration": {
    "ssl_cert_key": [
      "/usr/local/kong/ssl/kong-default.key",
      "/usr/local/kong/ssl/kong-default-ecdsa.key"
    ],
    "nginx_http_lua_regex_match_limit": "100000",
    "pluginserver_names": {},
    "pg_host": "127.0.0.1",
    "pg_database": "kong",
    "go_plugins_dir": "off",
    "pg_user": "kong",
    "go_pluginserver_exe": "/usr/local/bin/go-pluginserver",
    "nginx_proxy_real_ip_header": "X-Real-IP",
    "real_ip_recursive": "off",
    "nginx_proxy_real_ip_recursive": "off",
    "upstream_keepalive_pool_size": 60,
    "nginx_http_client_max_body_size": "0",
    "client_body_buffer_size": "8k",
    "nginx_http_client_body_buffer_size": "8k",
    "pg_port": 5432,
    "pg_ssl": false,
    "pg_ssl_verify": false,
    "pg_max_concurrent_queries": 0,
    "pg_semaphore_timeout": 60000,
    "pg_ro_ssl": false,
    "pg_ro_ssl_verify": false,
    "upstream_keepalive_idle_timeout": 60,
    "cassandra_contact_points": [
      "127.0.0.1"
    ],
    "ssl_cipher_suite": "intermediate",
    "cassandra_ssl": false,
    "cassandra_ssl_verify": false,
    "cassandra_write_consistency": "ONE",
    "cassandra_username": "kong",
    "lua_package_cpath": "",
    "error_default_type": "text/plain",
    "cassandra_lb_policy": "RequestRoundRobin",
    "mem_cache_size": "128m",
    "cassandra_refresh_frequency": 60,
    "cassandra_repl_strategy": "SimpleStrategy",
    "cassandra_repl_factor": 1,
    "cassandra_data_centers": [
      "dc1:2",
      "dc2:3"
    ],
    "cassandra_schema_consensus_timeout": 10000,
    "ssl_protocols": "TLSv1.1 TLSv1.2 TLSv1.3",
    "nginx_http_ssl_protocols": "TLSv1.2 TLSv1.3",
    "nginx_stream_ssl_protocols": "TLSv1.2 TLSv1.3",
    "ssl_prefer_server_ciphers": "on",
    "nginx_http_ssl_prefer_server_ciphers": "off",
    "nginx_stream_ssl_prefer_server_ciphers": "off",
    "ssl_dhparam": "ffdhe2048",
    "nginx_http_ssl_dhparam": "ffdhe2048",
    "nginx_stream_ssl_dhparam": "ffdhe2048",
    "ssl_session_tickets": "on",
    "db_update_frequency": 5,
    "nginx_stream_ssl_session_tickets": "on",
    "db_update_propagation": 0,
    "nginx_http_ssl_session_timeout": "1d",
    "db_cache_ttl": 0,
    "proxy_access_log": "/dev/stdout access_fmt",
    "proxy_error_log": "/dev/stderr",
    "proxy_stream_access_log": "logs/access.log basic",
    "proxy_stream_error_log": "logs/error.log",
    "admin_access_log": "/dev/stdout",
    "admin_error_log": "/dev/stderr",
    "status_access_log": "off",
    "status_error_log": "logs/status_error.log",
    "log_level": "notice",
    "nginx_status_directives": {},
    "nginx_optimizations": true,
    "lua_ssl_trusted_certificate": {},
    "lua_ssl_verify_depth": 1,
    "lua_ssl_protocols": "TLSv1.1 TLSv1.2 TLSv1.3",
    "nginx_http_lua_ssl_protocols": "TLSv1.1 TLSv1.2 TLSv1.3",
    "nginx_stream_lua_ssl_protocols": "TLSv1.1 TLSv1.2 TLSv1.3",
    "lua_socket_pool_size": 30,
    "cluster_control_plane": "127.0.0.1:8005",
    "kic": false,
    "nginx_http_upstream_directives": {},
    "cluster_mtls": "shared",
    "proxy_listeners": [
      {
        "port": 8000,
        "listener": "0.0.0.0:8000",
        "proxy_protocol": false,
        "deferred": false,
        "reuseport": false,
        "backlog=%d+": false,
        "bind": false,
        "ssl": false,
        "ip": "0.0.0.0",
        "http2": false
      },
      {
        "port": 8443,
        "listener": "0.0.0.0:8443 ssl http2",
        "proxy_protocol": false,
        "deferred": false,
        "reuseport": false,
        "backlog=%d+": false,
        "bind": false,
        "ssl": true,
        "ip": "0.0.0.0",
        "http2": true
      }
    ],
    "dns_resolver": {},
    "dns_hostsfile": "/etc/hosts",
    "untrusted_lua": "sandbox",
    "untrusted_lua_sandbox_requires": [
      "resty.sha256",
      "resty.string"
    ],
    "untrusted_lua_sandbox_environment": {},
    "dns_stale_ttl": 4,
    "dns_order": [
      "LAST",
      "SRV",
      "A",
      "CNAME"
    ],
    "dns_no_sync": false,
    "nginx_conf": "/usr/local/kong/nginx.conf",
    "worker_consistency": "strict",
    "nginx_kong_conf": "/usr/local/kong/nginx-kong.conf",
    "nginx_kong_stream_conf": "/usr/local/kong/nginx-kong-stream.conf",
    "kong_env": "/usr/local/kong/.kong_env",
    "ssl_cert_csr_default": "/usr/local/kong/ssl/kong-default.csr",
    "ssl_cert_default": "/usr/local/kong/ssl/kong-default.crt",
    "ssl_cert_key_default": "/usr/local/kong/ssl/kong-default.key",
    "ssl_cert_default_ecdsa": "/usr/local/kong/ssl/kong-default-ecdsa.crt",
    "ssl_cert_key_default_ecdsa": "/usr/local/kong/ssl/kong-default-ecdsa.key",
    "client_ssl_cert_default": "/usr/local/kong/ssl/kong-default.crt",
    "client_ssl_cert_key_default": "/usr/local/kong/ssl/kong-default.key",
    "admin_ssl_cert_default": "/usr/local/kong/ssl/admin-kong-default.crt",
    "admin_ssl_cert_key_default": "/usr/local/kong/ssl/admin-kong-default.key",
    "stream_proxy_ssl_enabled": false,
    "admin_ssl_enabled": true,
    "status_ssl_enabled": false,
    "admin_ssl_cert_key_default_ecdsa": "/usr/local/kong/ssl/admin-kong-default-ecdsa.key",
    "status_ssl_cert_default": "/usr/local/kong/ssl/status-kong-default.crt",
    "nginx_http_log_format": "access_fmt '$remote_addr - $remote_user [$time_iso8601] $host \"$request\" $status $body_bytes_sent $request_time \"$http_referer\" \"$http_user_agent\"'",
    "status_ssl_cert_default_ecdsa": "/usr/local/kong/ssl/status-kong-default-ecdsa.crt",
    "status_ssl_cert_key_default_ecdsa": "/usr/local/kong/ssl/status-kong-default-ecdsa.key",
    "nginx_http_lua_shared_dict": "prometheus_metrics 10m",
    "nginx_http_proxy_max_temp_file_size": "4096m",
    "port_maps": [
      "80:8000",
      "443:8443"
    ],
    "proxy_listen": [
      "0.0.0.0:8000",
      "0.0.0.0:8443 ssl http2"
    ],
    "admin_listen": [
      "127.0.0.1:8444 ssl"
    ],
    "status_listen": [
      "0.0.0.0:8100"
    ],
    "stream_listen": [
      "off"
    ],
    "cluster_listen": [
      "0.0.0.0:8005"
    ],
    "admin_ssl_cert": [
      "/usr/local/kong/ssl/admin-kong-default.crt",
      "/usr/local/kong/ssl/admin-kong-default-ecdsa.crt"
    ],
    "admin_ssl_cert_key": [
      "/usr/local/kong/ssl/admin-kong-default.key",
      "/usr/local/kong/ssl/admin-kong-default-ecdsa.key"
    ],
    "role": "traditional",
    "status_ssl_cert_key": {},
    "db_resurrect_ttl": 30,
    "anonymous_reports": true,
    "nginx_main_user": "kong kong",
    "nginx_daemon": "off",
    "nginx_main_daemon": "off",
    "nginx_worker_processes": "2",
    "nginx_main_worker_processes": "2",
    "cassandra_timeout": 5000,
    "trusted_ips": {},
    "real_ip_header": "X-Real-IP",
    "worker_state_update_frequency": 5,
    "client_max_body_size": "0",
    "cassandra_port": 9042,
    "status_ssl_cert": {},
    "host_ports": {
      "8443": 443,
      "8000": 80
    },
    "admin_ssl_cert_default_ecdsa": "/usr/local/kong/ssl/admin-kong-default-ecdsa.crt",
    "cassandra_read_consistency": "ONE",
    "client_ssl": false,
    "nginx_proxy_directives": [
      {
        "name": "real_ip_header",
        "value": "X-Real-IP"
      },
      {
        "name": "real_ip_recursive",
        "value": "off"
      }
    ],
    "db_cache_warmup_entities": [
      "services"
    ],
    "nginx_http_ssl_session_tickets": "on",
    "ssl_session_timeout": "1d",
    "nginx_stream_ssl_session_timeout": "1d",
    "database": "off",
    "lua_package_path": "/opt/?.lua;;",
    "headers": [
      "server_tokens",
      "latency_tokens"
    ],
    "prefix": "/usr/local/kong",
    "status_listeners": [
      {
        "port": 8100,
        "ssl": false,
        "ip": "0.0.0.0",
        "listener": "0.0.0.0:8100"
      }
    ],
    "cluster_data_plane_purge_delay": 1209600,
    "cluster_ocsp": "off",
    "dns_error_ttl": 1,
    "dns_not_found_ttl": 30,
    "cluster_listeners": [
      {
        "port": 8005,
        "listener": "0.0.0.0:8005",
        "proxy_protocol": false,
        "deferred": false,
        "reuseport": false,
        "backlog=%d+": false,
        "bind": false,
        "ssl": false,
        "ip": "0.0.0.0",
        "http2": false
      }
    ],
    "ssl_cert": [
      "/usr/local/kong/ssl/kong-default.crt",
      "/usr/local/kong/ssl/kong-default-ecdsa.crt"
    ],
    "cluster_v2": false,
    "nginx_main_worker_rlimit_nofile": "auto",
    "plugins": [
      "bundled",
      "hasher"
    ],
    "nginx_events_worker_connections": "auto",
    "ssl_ciphers": "ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384",
    "nginx_events_multi_accept": "on",
    "pg_timeout": 5000,
    "nginx_user": "kong kong",
    "upstream_keepalive_max_requests": 100,
    "cassandra_keyspace": "kong",
    "nginx_main_directives": [
      {
        "name": "daemon",
        "value": "off"
      },
      {
        "name": "user",
        "value": "kong kong"
      },
      {
        "name": "worker_processes",
        "value": "2"
      },
      {
        "name": "worker_rlimit_nofile",
        "value": "auto"
      }
    ],
    "nginx_events_directives": [
      {
        "name": "multi_accept",
        "value": "on"
      },
      {
        "name": "worker_connections",
        "value": "auto"
      }
    ],
    "nginx_http_directives": [
      {
        "name": "client_body_buffer_size",
        "value": "8k"
      },
      {
        "name": "client_max_body_size",
        "value": "0"
      },
      {
        "name": "log_format",
        "value": "access_fmt '$remote_addr - $remote_user [$time_iso8601] $host \"$request\" $status $body_bytes_sent $request_time \"$http_referer\" \"$http_user_agent\"'"
      },
      {
        "name": "lua_regex_match_limit",
        "value": "100000"
      },
      {
        "name": "lua_shared_dict",
        "value": "prometheus_metrics 10m"
      },
      {
        "name": "lua_ssl_protocols",
        "value": "TLSv1.1 TLSv1.2 TLSv1.3"
      },
      {
        "name": "proxy_max_temp_file_size",
        "value": "4096m"
      },
      {
        "name": "ssl_dhparam",
        "value": "/usr/local/kong/ssl/ffdhe2048.pem"
      },
      {
        "name": "ssl_prefer_server_ciphers",
        "value": "off"
      },
      {
        "name": "ssl_protocols",
        "value": "TLSv1.2 TLSv1.3"
      },
      {
        "name": "ssl_session_tickets",
        "value": "on"
      },
      {
        "name": "ssl_session_timeout",
        "value": "1d"
      }
    ],
    "nginx_upstream_directives": {},
    "nginx_admin_directives": [
      {
        "name": "client_body_buffer_size",
        "value": "10m"
      },
      {
        "name": "client_max_body_size",
        "value": "10m"
      }
    ],
    "nginx_stream_directives": [
      {
        "name": "lua_shared_dict",
        "value": "stream_prometheus_metrics 5m"
      },
      {
        "name": "lua_ssl_protocols",
        "value": "TLSv1.1 TLSv1.2 TLSv1.3"
      },
      {
        "name": "ssl_dhparam",
        "value": "/usr/local/kong/ssl/ffdhe2048.pem"
      },
      {
        "name": "ssl_prefer_server_ciphers",
        "value": "off"
      },
      {
        "name": "ssl_protocols",
        "value": "TLSv1.2 TLSv1.3"
      },
      {
        "name": "ssl_session_tickets",
        "value": "on"
      },
      {
        "name": "ssl_session_timeout",
        "value": "1d"
      }
    ],
    "nginx_supstream_directives": {},
    "nginx_sproxy_directives": {},
    "loaded_plugins": {
      "jwt": true,
      "acl": true,
      "correlation-id": true,
      "cors": true,
      "oauth2": true,
      "tcp-log": true,
      "udp-log": true,
      "file-log": true,
      "http-log": true,
      "key-auth": true,
      "hmac-auth": true,
      "basic-auth": true,
      "ip-restriction": true,
      "request-transformer": true,
      "response-transformer": true,
      "request-size-limiting": true,
      "rate-limiting": true,
      "response-ratelimiting": true,
      "syslog": true,
      "loggly": true,
      "datadog": true,
      "ldap-auth": true,
      "statsd": true,
      "bot-detection": true,
      "aws-lambda": true,
      "request-termination": true,
      "prometheus": true,
      "proxy-cache": true,
      "session": true,
      "acme": true,
      "grpc-gateway": true,
      "grpc-web": true,
      "pre-function": true,
      "post-function": true,
      "azure-functions": true,
      "zipkin": true,
      "hasher": true
    },
    "proxy_ssl_enabled": true,
    "stream_listeners": {},
    "admin_listeners": [
      {
        "port": 8444,
        "listener": "127.0.0.1:8444 ssl",
        "proxy_protocol": false,
        "deferred": false,
        "reuseport": false,
        "backlog=%d+": false,
        "bind": false,
        "ssl": true,
        "ip": "127.0.0.1",
        "http2": false
      }
    ],
    "nginx_http_status_directives": {},
    "nginx_admin_client_max_body_size": "10m",
    "enabled_headers": {
      "X-Kong-Proxy-Latency": true,
      "X-Kong-Response-Latency": true,
      "X-Kong-Admin-Latency": true,
      "server_tokens": true,
      "X-Kong-Upstream-Latency": true,
      "X-Kong-Upstream-Status": false,
      "Via": true,
      "latency_tokens": true,
      "Server": true
    },
    "nginx_pid": "/usr/local/kong/pids/nginx.pid",
    "nginx_admin_client_body_buffer_size": "10m",
    "nginx_err_logs": "/usr/local/kong/logs/error.log",
    "nginx_acc_logs": "/usr/local/kong/logs/access.log",
    "status_ssl_cert_key_default": "/usr/local/kong/ssl/status-kong-default.key",
    "admin_acc_logs": "/usr/local/kong/logs/admin_access.log"
  },
  "timers": {
    "pending": 11,
    "running": 89
  },
  "tagline": "Welcome to kong",
  "pids": {
    "master": 1,
    "workers": [
      1097,
      1098
    ]
  },
  "version": "2.6.0",
  "hostname": "ingress-kong-85ff46d4d7-vhk75",
  "node_id": "d2aa8190-c85b-4b3a-8de0-dfc72ba1e5ec",
  "plugins": {
    "enabled_in_cluster": [
      "hasher",
      "basic-auth",
      "zipkin",
      "acl",
      "prometheus",
      "ip-restriction"
    ],
    "available_on_server": {
      "jwt": true,
      "acl": true,
      "correlation-id": true,
      "cors": true,
      "oauth2": true,
      "tcp-log": true,
      "udp-log": true,
      "file-log": true,
      "http-log": true,
      "key-auth": true,
      "hmac-auth": true,
      "basic-auth": true,
      "ip-restriction": true,
      "request-transformer": true,
      "response-transformer": true,
      "request-size-limiting": true,
      "rate-limiting": true,
      "response-ratelimiting": true,
      "syslog": true,
      "loggly": true,
      "datadog": true,
      "ldap-auth": true,
      "statsd": true,
      "bot-detection": true,
      "aws-lambda": true,
      "request-termination": true,
      "prometheus": true,
      "proxy-cache": true,
      "session": true,
      "acme": true,
      "grpc-gateway": true,
      "grpc-web": true,
      "pre-function": true,
      "post-function": true,
      "azure-functions": true,
      "zipkin": true,
      "hasher": true
    }
  },
  "lua_version": "LuaJIT 2.1.0-beta3"
}
```
</details>

**Comment by seh** [comment]
We had a difficult afternoon and evening yesterday, as this problem overwhelmed us again, this time in our GKE-hosted Kubernetes cluster in GCP.

Once again, we started out with three Kong replicas serving a low rate of requests—around 125 RPS—with their proxy containers' CPU usage low at around 1/3 of a CPU. A surge of requests arrived, almost instantly overwhelming these proxies. They started responding with HTTP status code 404 to all requests. Once again, Kong's Prometheus plugin does not capture these responses. For some reason, it seems that the plugin is blind to the 404 responses that occur when in this state. We _did_ see a brief rise in both status code 200 and 101 responses before they fell down to zero. These 101 responses are likely due to WebSocket connections.

We have a _HorizontalPodAutoscaler_ watching Kong's _Deployment_. It didn't add any replicas, because these pods' CPU usage fell too low—after a brief increase—to cross the threshold for adding more replicas. Recognizing that these three pods were "stuck," thinking of what we learned the day prior, we deleted two of the existing three pods, and orphaned one of them from the _Deployment_ to keep it around for inspection, outside of the set selected by the _Service_. This time, though, the replacement pods **did not start serving normally**. They started, and immediately started serving all responses with the 404 status code.

We forced our _HorizontalPodAutoscaler_ to increase the replica count to ten pods, and only as seven more pods started did we start seeing some of the expected 200 status codes in the responses. Several of these new pods then fell ill as well, getting "stuck" serving all responses with the 404 status code. We added yet more pods, until we had thirty of them. Shortly afterward we noticed something we had not seen before: **a few of the pods that were "stuck" serving all 404s recovered on their own.**

Looking through the proxy container's log in the orphaned pod that I saved—one of the initial three—I can see that it logging each of the responses with 404 status code that it served, and _only_ logging those. Once I took it out of the _Service_'s selected set, it started logging its repeated internal health checks. Here is that transition:
<details>
<summary>Sample of proxy log messages</summary>

```
15.177.46.82 - - [2021-12-07T21:13:02+00:00] 3938446c-cbe6-4336-9309-6beb6fa4f63c.example.com "GET /health HTTP/1.1" 404 48 0.000 "-" "Amazon-Route53-Health-Check-Service (ref 6bd0a561-dd95-43c2-aded-b690ef719847; report http://amzn.to/1vsZADi)"
15.177.42.96 - - [2021-12-07T21:13:02+00:00] 01350754-4c57-44b9-b1ee-c68c877865ab.example.com "GET /health HTTP/1.1" 404 48 0.000 "-" "Amazon-Route53-Health-Check-Service (ref 8d3f23c0-af32-4748-891f-fbfe1c89f030; report http://amzn.to/1vsZADi)"
15.177.62.72 - - [2021-12-07T21:13:02+00:00] primary.8697afd8-a7ad-469e-b4dd-f099cf8ca744.example.com "GET /health HTTP/1.1" 404 48 0.000 "-" "Amazon-Route53-Health-Check-Service (ref 1eb31b46-5a64-44fe-b554-d22454326296; report http://amzn.to/1vsZADi)"
2021/12/07 21:29:56 [notice] 1096#0: *1827908 [lua] init.lua:260: purge(): [DB cache] purging (local) cache, client: 127.0.0.1, server: kong_admin, request: "POST /config?check_hash=1 HTTP/1.1", host: "127.0.0.1:8444"
2021/12/07 21:29:56 [notice] 1096#0: *1827908 [lua] init.lua:260: purge(): [DB cache] purging (local) cache, client: 127.0.0.1, server: kong_admin, request: "POST /config?check_hash=1 HTTP/1.1", host: "127.0.0.1:8444"
127.0.0.1 - - [07/Dec/2021:21:29:56 +0000] "POST /config?check_hash=1 HTTP/1.1" 201 229139 "-" "Go-http-client/1.1"
2021/12/07 21:30:06 [warn] 1096#0: *1828709 [lua] healthcheck.lua:1126: log(): [healthcheck] (0dc6f45b-8f8d-40d2-a504-473544ee190b:some-svc.834c39c9-d605-4c78-9f2b-21938197fee6.8888.svc) unhealthy HTTP increment (1/5) for '10.33.238.7(10.33.238.7:8888)', context: ngx.timer, client: 127.0.0.1, server: 127.0.0.1:8444
2021/12/07 21:30:06 [warn] 1096#0: *1828709 [lua] healthcheck.lua:1126: log(): [healthcheck] (0dc6f45b-8f8d-40d2-a504-473544ee190b:some-svc.834c39c9-d605-4c78-9f2b-21938197fee6.8888.svc) unhealthy HTTP increment (1/5) for '10.33.239.7(10.33.239.7:8888)', context: ngx.timer, client: 127.0.0.1, server: 127.0.0.1:8444
```
</details>

Today I inspected some of the Admin API output from this orphaned pod.
<details>
<summary>"/status" path</summary>

```json
{
  "server": {
    "connections_writing": 1,
    "connections_accepted": 919779,
    "connections_waiting": 0,
    "connections_handled": 919779,
    "total_requests": 48473,
    "connections_active": 1,
    "connections_reading": 0
  },
  "database": {
    "reachable": true
  },
  "memory": {
    "workers_lua_vms": [
      {
        "pid": 1096,
        "http_allocated_gc": "84.03 MiB"
      },
      {
        "pid": 1097,
        "http_allocated_gc": "138.38 MiB"
      }
    ],
    "lua_shared_dicts": {
      "kong_process_events": {
        "capacity": "5.00 MiB",
        "allocated_slabs": "5.00 MiB"
      },
      "kong_cluster_events": {
        "capacity": "5.00 MiB",
        "allocated_slabs": "0.04 MiB"
      },
      "kong_rate_limiting_counters": {
        "capacity": "12.00 MiB",
        "allocated_slabs": "0.08 MiB"
      },
      "kong_core_db_cache": {
        "capacity": "128.00 MiB",
        "allocated_slabs": "18.60 MiB"
      },
      "kong_core_db_cache_miss": {
        "capacity": "12.00 MiB",
        "allocated_slabs": "0.09 MiB"
      },
      "kong_db_cache": {
        "capacity": "128.00 MiB",
        "allocated_slabs": "0.76 MiB"
      },
      "kong_db_cache_miss": {
        "capacity": "12.00 MiB",
        "allocated_slabs": "0.08 MiB"
      },
      "kong_core_db_cache_2": {
        "capacity": "128.00 MiB",
        "allocated_slabs": "18.61 MiB"
      },
      "kong_core_db_cache_miss_2": {
        "capacity": "12.00 MiB",
        "allocated_slabs": "0.08 MiB"
      },
      "kong_db_cache_2": {
        "capacity": "128.00 MiB",
        "allocated_slabs": "0.76 MiB"
      },
      "kong_db_cache_miss_2": {
        "capacity": "12.00 MiB",
        "allocated_slabs": "0.08 MiB"
      },
      "prometheus_metrics": {
        "capacity": "10.00 MiB",
        "allocated_slabs": "0.08 MiB"
      },
      "kong": {
        "capacity": "5.00 MiB",
        "allocated_slabs": "0.04 MiB"
      },
      "kong_locks": {
        "capacity": "8.00 MiB",
        "allocated_slabs": "0.06 MiB"
      },
      "kong_healthchecks": {
        "capacity": "5.00 MiB",
        "allocated_slabs": "0.50 MiB"
      }
    }
  }
}
```
</details>

The "/routes" path output includes routes for all of our _Ingress_ objects. The "/services" path output includes services for all of our upstream _Service_ objects used by the _Ingress_ objects. This tells me that the ingress controller has been able to push this configuration to the proxy since it got "stuck."

Seeing this, I relabeled this pod to allow it to be selected by Kong's  _Service_. It started serving responses with status code 200. That tells me again that getting "stuck" like this is not a permanent condition. It seems to persist so long as the request load persists.

What else can we tell you to help figure out what is wrong with Kong here?

**Comment by seh** [comment]
As part of this same incident, we suffered a problem with different symptoms. As a surge of clients arrived starting WebSocket connections, we had a set of 32 upstream servers (for a particular _Ingress_ and _Service_ pair) that were having trouble serving the connections. Kong's Prometheus metrics showed a huge surge in responses with status code 101 as all the 200 status code-bearing responses tapered off.

Each of the Kong proxy containers' CPU usage rose to two whole CPUs—equal to our resource request in the pod spec—and the _HorizontalPodAutoscaler_ recognized this 95-100% CPU usage range. We allowed our _HPA_ to create as many as 50 Kong pods. **All of them** were using two CPUs. During this period, Kong reported around 250 total connections across all fifty pods. It seems that something the upstream servers were doing—and doing poorly—was causing Kong to burn a lot of CPU. It's hard to tell what's going on with these WebSocket connections, but some of our customers assured us that their clients were not communicating successfully during this period.

**Comment by hbagdi** [comment]
@dndx @bungle Have we heard of cases where Kong's router stops working altogether?

**Comment by bungle** [comment]
@hbagdi, I haven't but I have seen some workers being killed with host on router rebuilds with huge configs (70k+ routes). Right now I am lost what causes this. The installation seems to be `dbless` Kong proxies right? Are there changes to config during this surge (e.g. does Kong need to rebuild router during surge)?

**Comment by seh** [comment]
Yes, we are not using a database for the proxies. During this surge, we don't expect there would be any configuration changes. We might have a replica come or go from the upstream pods (changing the Kong service), but none of these upstream pods are controlled by a _HorizontalPodAutoscaler_, so their population is not governed by incident request load.

As for plugins, we use the following:
- _acl_
- _basic-auth_
- _cors_  
  (used sporadically)
- _hasher_  
  (written in house, rewrites the HTTP "Authorization" header with the SHA-256 sum of the password value)
- _prometheus_  
  (active globally)
- _zipkin_
  (used sporadically for now)

**Comment by seh** [comment]
> Once again, Kong's Prometheus plugin does not capture these responses. For some reason, it seems that the plugin is blind to the 404 responses that occur when in this state.

I think I figured out why this is so. Is it correct that [the branch that handles the lack of a router match](https://sourcegraph.com/github.com/Kong/kong@master/-/blob/kong/runloop/handler.lua?L1217-L1221) completes [before Kong runs any plugins](https://sourcegraph.com/github.com/Kong/kong@master/-/blob/kong/runloop/handler.lua?L1010-L1011)? That branch lies within the "before" field (phase?) of the "init_worker" phase.

If that's so, is there any way other than the access log that Kong could expose this behavior? The user's intention when activating the Prometheus plugin _globally_ is to observe **all** of Kong's HTTP responses, among other things.

**Comment by dndx** [comment]
Hello @seh,

You are indeed correct. Kong's entire plugin system is driven by match result of the router, so without a router match it is not really safe to run the plugin (e.g. some context value such as match information will be missing for the plugin).

I am wondering could the 404 be caused by Kong proxy being restarted and could not get the config in time from KIC? Could you check the output of `GET /config` on DP when 404 happens and does it output the correct config?

**Comment by seh** [comment]
> Kong's entire plugin system is driven by match result of the router, so without a router match it is not really safe to run the plugin (e.g. some context value such as match information will be missing for the plugin).

That makes sense, but we're wondering whether it's feasible to count these route lookup misses and export them in a Prometheus metric some other way, separate from the Prometheus plugin. Trying to operate Kong and not being able to observe these cases where Kong responds with a 404 status code—apart from log inspection—is dangerous.

> I am wondering could the 404 be caused by Kong proxy being restarted and could not get the config in time from KIC?

We have seen the kubelet kill the proxy container when it fails its liveness probe enough times consecutively. It's failing that probe either by not accepting the HTTP connection or by not responding quickly enough, which suggests that [Kong can't fetch NGINX's _/nginx_status_ HTTP URL](https://sourcegraph.com/github.com/Kong/kong@master/-/blob/kong/api/routes/health.lua?L42). However, [Kong's ingress controller won't declare itself ready until the ingress controller has pushed the configuration in at least once](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller@994fa04/-/blob/internal/proxy/clientgo_cached_proxy_resolver.go?L165-L176). That keeps the pod's IP addresses out of the _Endpoints_ and _EndpointSlice_ objects' ready set.

> Could you check the output of `GET /config` on DP when 404 happens and does it output the correct config?

Next time we can catch it in this state, yes, I'll try. The output that [I included above](https://github.com/Kong/kong/issues/8152#issue-1073567416) is from one of the containers that got into this state, but then got itself out of this state by the time I could issue the HTTP request against it.

**Comment by seh** [comment]
> However, [Kong's ingress controller won't declare itself ready until the ingress controller has pushed the configuration in at least once](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller@994fa04/-/blob/internal/proxy/clientgo_cached_proxy_resolver.go?L165-L176). That keeps the pod's IP addresses out of the _Endpoints_ and _EndpointSlice_ objects' ready set.

Having thought about this more today, and inspecting Kong's ingress controller code, I found what I think is a problem: **If the proxy container gets killed after the ingress controller has pushed configuration to the proxy at least once before, assuming "reverse sync" is disabled, [the ingress controller will not push the configuration again into any replacement proxy containers until the observed cluster state changes](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller@994fa04/-/blob/internal/sendconfig/sendconfig.go?L47-57).** On top of that, the ingress controller's readiness gate is latching: [once set to true, it never returns to false](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller@994fa0404344e97a014366789a3f349dfd3d812d/-/blob/internal/proxy/clientgo_cached_proxy_resolver.go?L276:4&subtree=true).

This means that it's possible for the ingress controller and proxy containers to start, synchronize configuration, both become ready, and start serving requests. Then the proxy container dies, and a replacement container comes up, but the ingress controller doesn't push any configuration to the proxy container. It _considers doing so_ once every [three seconds by default](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller@994fa04/-/blob/internal/proxy/proxy.go?L29-L37)—the so-called "stagger" duration—but it optimizes away the attempt if the intended configuration has not changed [since the last time it applied it to the proxy](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller@994fa04/-/blob/internal/sendconfig/sendconfig.go?L94-L95). The proxy container reports that it's ready, and the ingress controller container continues to report that it's ready. The pod's IP address remains in the _Endpoints_ object's ready set.

This doesn't speak to what leads the proxy container's liveness probe to fail such that the kubelet kills it, but after it's been killed and replaced, if the observed cluster state (_Ingress_, _Service_, _Endpoints_, and _EndpointSlice_ objects) don't churn, the proxy can sit without any material configuration, rejecting all requests for lack of a matching route.

**Comment by shaneutt** [comment]
@seh thank you very much for the in-depth analysis. I work on @Kong/team-k8s and we appreciate you bringing what appears to be a second and consequent issue to light. Assuming this current issue is going to lead to a fix in the gateway it sounds like doing so might eventually "hide" the consequent issue in KIC that you've described, which we'll want to avoid.

We should make a separate issue in [kong/kubernetes-ingress-controller](https://github.com/kong/kubernetes-ingress-controller) for this problem and link it here for reference. I am happy to create that issue myself if you like, but wanted to ask since you've done the work if you would prefer to make that issue? Let me know I'm OK with either :vulcan_salute:

**Comment by seh** [comment]
All is well, Shane. I'll file an issue against the ingress controller and link it here. I'll be more succinct there than usual, given that we're already underway here.

**Comment by seh** [comment]
Please see Kong/kubernetes-ingress-controller#2107.

**Comment by donalddewulf** [comment]
Any news about this? It's especially problematic during a scale up.

**Comment by jcam** [comment]
We got hit with this yesterday on our production workloads as well. Deleting the failed pod so it reinitializes fully resolves the issue. Still investigating why the proxy quit in the first place.

**Comment by kollol-chowdhury-incontact** [comment]
We are facing almost the same issue with kong 2.7.1. After 5mins of starting kong,it starts giving 404 error./once kong reload is done,it works again for like 3-4mins and again starts giving 404 error. We are running dockerized kong in ec2. There is no logs to understand what is happening even. Only thing which i found is lots of out of memory/Lua VM crashed during that timeframe.

**Comment by seh** [comment]
I'm not sure of the current state of the proxy on this front. Are there any outstanding patches under review?
